scripts/end_to_end_test/run_end_to_end_test.py

```diff
@@ -550,7 +570,93 @@ def wait_network_ready(parameters, test_instance):
         f"Network readiness check failed, because reached max trials ({max_trials})")
 
 
+def query_balance(parameters, test_instance):
+    nodes = parameters["nodes"]
+    variable = parameters["save_as"]
+    for node_index in nodes:
+        node_instance = test_instance._nodes[node_index]
+        node_host = "localhost"
+        node_port = node_instance._port_start
+
+        api = LedgerApi(node_host, node_port)
+
+        # create the entity from the node's private key
+        entity = Entity(get_nodes_private_key(test_instance, node_index))
+        b = api.tokens.balance(entity)
+        address = Address(entity)
+        output(
+            f"Requested balance ({b}) {address.to_hex()}, saved as {variable}")
+        if not hasattr(node_instance, "_variables"):
+            setattr(node_instance, "_variables", {})
+        node_instance._variables[variable] = b
+
+
+def execute_expression(parameters, test_instance):
```

Comment:
This function might be a good candidate to use `eval()`: https://docs.python.org/3/library/functions.html#eval

Reply:
I'm aware of eval, it was a choose not using it, to only allow the predefined expressions. And also the variable names are simple names in the expression but in reality they are attached to test instance.